Handle some numpy functions with out arguments correctly in dynamo

jamesjwu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #118248


Dynamo creates Tensors when tracing through numpy ufuncs like np.sin, np.minimum etc. When running, np functions generally return Tensors when run with `torch.compile`. However, we currently require when normalizing `out` arguments that the input is an ndarray.  This creates assertion errors when running torch.compile on any numpy function with an out argument: 
```
    def test_numpy_ufunc_out(self):
        @torch.compile(backend="eager")
        def foo():
            x = np.arange(5)
            out = np.empty((x.shape[0], x.shape[0]))
            res_out = np.sin(x, out=out)
            assert res_out is out
        foo() 
```
Failure with stack trace: https://gist.github.com/jamesjwu/68e217638d735678b3de968584dba23f

Instead, we can wrap tensors in an ndarray in normalize_outarray to handle the case correctly. Fixing this resolves ~220 tests under dynamo_test_failures, but also exposes a followup bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

jamesjwu committed:
Wrap tensor arguments in normalize_outarray properly

[ghstack-poisoned]

jamesjwu committed:
Update on "Wrap tensor arguments in normalize_outarray properly"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 

Fixing this resolves ~200 tests under dynamo_test_failures, but it also exposes a completely separate bug. 

In the presence of a graph break, ndarrays no longer preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```




[ghstack-poisoned]

jamesjwu committed:
Update on "Wrap tensor arguments in normalize_outarray properly"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 

Fixing this resolves ~200 tests under dynamo_test_failures, but it also exposes a completely separate bug. 

In the presence of a graph break, ndarrays no longer preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu commented on torch/testing/_internal/dynamo_test_failures.py:
```diff
@@ -239,229 +239,25 @@
-    "TestBinaryUfuncs.test_xy_and_out_casting_casting_unsafe_ufunc5_out_dtype_complex128",
-    "TestBinaryUfuncs.test_xy_and_out_casting_casting_unsafe_ufunc4_out_dtype_complex128",
-    "TestBinaryUfuncs.test_xy_and_out_casting_casting_equiv_ufunc6_out_dtype_complex128",
+    "TestBinaryUfuncs.test_xy_and_out_broadcast_ufunc16",
```
This is actually previously on line 254, will move back

jamesjwu committed:
Update on "Wrap tensor arguments in normalize_outarray properly"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu committed:
Update on "Wrap FakeTensor arguments in normalize_outarray in ndarrays during dynamo tracing"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano commented:
I see the issue. I don't think this is the best possible fix tho. Let me think about it a bit tomorrow.

jamesjwu commented:
Actually, how about putting it in the numpy_to_tensor_wrapper that we have in dynamo already? 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/utils.py#L1996

That way it only affects dynamo (though it still doesn't yet allow us to trace across graph breaks properly)

jamesjwu committed:
Update on "Wrap FakeTensor arguments in normalize_outarray in ndarrays during dynamo tracing"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano reviewed (commented):
So, I don't quite understand what's the issue that we are trying to solve here. The `out` parameter logic is handled here:
https://github.com/pytorch/pytorch/blob/ea851eb0271bb992ba9d4e62eae8085cc3bc2f95/torch/_numpy/_normalizations.py#L236-L242
and it seems correct to me. Can you write a standalone test that shows the issue you are trying to solve?

I don't know why do we have all these skips in `dynamo_test_failures.py`. These files are already being run with dynamo enabled. We enabled them in https://github.com/pytorch/pytorch/pull/112141, https://github.com/pytorch/pytorch/pull/112146 and quite a few other PRs. We put skips as needed in the tests themselves. I don't think it's correct to rerun them in `dynamo_test_failures.py`.

zou3519 commented:
> don't know why do we have all these skips in dynamo_test_failures.py. These files are already being run with dynamo enabled. We enabled them in https://github.com/pytorch/pytorch/pull/112141, https://github.com/pytorch/pytorch/pull/112146 and quite a few other PRs. We put skips as needed in the tests themselves. I don't think it's correct to rerun them in dynamo_test_failures.py.

@lezcano, many of the test/torch_np tests fail when run with PYTORCH_TEST_WITH_DYNAMO=1.  The failure in this issue can be reproduced by:
- removing `TestUnaryUfuncs.test_x_and_out_broadcast_ufunc0` in dynamo_test_failures.py
- running `PYTORCH_TEST_WITH_DYNAMO=1 python test/torch_np/test_ufuncs_basic.py -k test_x_and_out_broadcast_ufunc0`

dynamo_test_failures.py auto-generates skip and expectedFailure decorators.

Now, here's the story behind `dynamo_test_failures.py`. Previously, `PYTORCH_TEST_WITH_DYNAMO=1 ` would run unittests with Dynamo, but with suppress_errors=True, which hid real errors. We recently did a run to change it so that `PYTORCH_TEST_WITH_DYNAMO=1 ` runs Dynamo with suppress_errors=False, which is closer to what happens when a user calls torch.compile. This led to us identifying a lot of failing tests, which we added to `dynamo_test_failures.py`. Sorry for the confusion -- we just finished constructing `dynamo_test_failures.py` and I hadn't had the time to communicate this yet.

jamesjwu commented:
This particular PR is attempting to fix an issue where np funcs with `out` arguments throw errors when using torch.compile. 
Here's a simple example without needing PYTORCH_TEST_WITH_DYNAMO as well:
```
    def test_numpy_ufunc_out(self):
        @torch.compile(backend="eager")
        def foo():
            x = np.arange(5)
            out = np.empty((x.shape[0], x.shape[0]))
            res_out = np.sin(x, out=out)
            assert res_out is out
        foo() 
```
Failure with stack trace: https://gist.github.com/jamesjwu/68e217638d735678b3de968584dba23f

lezcano commented:
Does `PYTORCH_TEST_WITH_DYNAMO`  set `TEST_WITH_TORCHDYNAMO=1`? We have so many of these
https://github.com/pytorch/pytorch/blob/fae569b4f2a77389015c62fb863c2d7b72e2e76d/test/torch_np/numpy_tests/lib/test_arraypad.py#L13-L20
in the tests.

FWIW, the fix should be somewhere in
https://github.com/pytorch/pytorch/blob/ea851eb0271bb992ba9d4e62eae8085cc3bc2f95/torch/_numpy/_normalizations.py#L236-L242
probably. We should just be a bit more careful with the logic there so that we preserve the identity of the out tensor.

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu commented:
Yup, it sets that here I believe: https://github.com/pytorch/pytorch/blame/476b744e23dc9bdf2066534dbab855d6257769de/torch/testing/_internal/common_utils.py#L1319

So I can try to add the fix there, but unfortunately the assertion happens earlier when we try to normalize the params to the function (line 205 or so in that file). To me, it makes sense that the fix live in the dynamo code, since it's a dynamo specific issue (and that way, we don't unnecessarily do it when a user passes a tensor to an out parameter in eager mode?), but I could also wrap the out arg before normalization in that function, if you prefer.

jamesjwu commented:
Oh, actually on rereading your comment maybe you mean a fix to handle preserving ids across graph breaks should be in _normalizations? That makes sense, will dig deeper on that one too (probably in a separate PR).

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu commented:
Chatted with @lezcano offline and updated PR to the original approach doing the wrapping in normalization_outarray

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates FakeTensors when tracing through numpy ufuncs like np.sin, np.minimum etc. During normalization, we usually expect that inputs to numpy functions are numpy arrays. However, because the outarray of a numpy operation is linked to the result of the numpy operation itself, we may pass a Tensor as an argument when tracing for example values of the results of various ufuncs. 

Therefore, we should handle that gracefully and wrap the Tensor in a ndarray when handling the out argument. 
To be clear, I'm not 100% sure if wrapping it in normalize_outarray is the best fix, but it does resolve the issue by allowing dynamo to finish tracing. 

Fixing this resolves ~220 tests under dynamo_test_failures, but it also exposes a separate bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano reviewed (approved):
Thank you for the fix!

lezcano commented:
@pytorchbot merge

jamesjwu commented:
@pytorchbot merge

jamesjwu committed:
Update on "Handle some numpy functions with out arguments correctly in dynamo"



Dynamo creates Tensors when tracing through numpy ufuncs like np.sin, np.minimum etc. When running, np functions generally return Tensors when run with `torch.compile`. However, we currently require when normalizing `out` arguments that the input is an ndarray.  This creates assertion errors when running torch.compile on any numpy function with an out argument: 
```
    def test_numpy_ufunc_out(self):
        torch.compile(backend="eager")
        def foo():
            x = np.arange(5)
            out = np.empty((x.shape[0], x.shape[0]))
            res_out = np.sin(x, out=out)
            assert res_out is out
        foo() 
```
Failure with stack trace: https://gist.github.com/jamesjwu/68e217638d735678b3de968584dba23f

Instead, we can wrap tensors in an ndarray in normalize_outarray to handle the case correctly. Fixing this resolves ~220 tests under dynamo_test_failures, but also exposes a followup bug. 

In the presence of a graph break, ndarrays don't preserve their id, which can affect assertions and `is` checks between numpy arrays:
```
     def test_x_and_out_broadcast(self, ufunc):
        x = self.get_x(ufunc)
        out = np.empty((x.shape[0], x.shape[0]))

        x_b = np.broadcast_to(x, out.shape)
        # ufunc is just np.sin here
        res_out = ufunc(x, out=out)
        res_bcast = ufunc(x_b)
        # passes
        assert res_out is out
        graph_break()
        # fails
        assert res_out is out
```
Regular tensors preserve their id because Dynamo caches their example tensor values across a graph break. However, with ndarrays, we only store their converted tensor values, and construct new ndarrays around those values: 
https://github.com/pytorch/pytorch/blob/eebe7e1d37f1baa995c694d540cc2fc98884fa18/torch/_dynamo/variables/builder.py#L1083
Added a test with expected failure to showcase this — we can then fix that issue separately.



cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
